//------------------------------------------------------------------------------
// {C} Copyright 2018 Pensando Systems Inc. All rights reserved
//
// protobuf specification for system level objects
//------------------------------------------------------------------------------

syntax = "proto3";

package device;

import "delphi.proto";
import "gogo.proto";

//------------------------------------------------------------------------------
// supported forwarding modes
//------------------------------------------------------------------------------
enum ForwardingMode {
  FORWARDING_MODE_NONE       = 0;
  FORWARDING_MODE_CLASSIC    = 1;    // classic forwarding
  FORWARDING_MODE_HOSTPIN    = 2;    // smartnic hostpin mode
  FORWARDING_MODE_SWITCH     = 3;    // smartnic switch mode
}

//------------------------------------------------------------------------------
// supported feature profiles
//------------------------------------------------------------------------------
enum FeatureProfile {
  FEATURE_PROFILE_NONE                  = 0;
  FEATURE_PROFILE_CLASSIC_DEFAULT       = 1;  // default classic mode profile with 2 host devices
  FEATURE_PROFILE_CLASSIC_ETH_DEV_SCALE = 2;  // scaled up classic mode w.r.t devices
}

enum PortAdminState {
  PORT_ADMIN_STATE_ENABLE   = 0;
  PORT_ADMIN_STATE_DISABLE  = 1;
}

enum MircoSegMode {
 MICRO_SEG_DISABLE = 0;
 MICRO_SEG_ENABLE  = 1;
}

//------------------------------------------------------------------------------
// SystemSpec captures the system level global configuraiton
// NOTE:
// fwd_mode is init time configuration and this object must be pushed
// by delphi and client handler should be invoked before any other
// config object is
//------------------------------------------------------------------------------
message SystemSpec {
  option (delphi.update_event)          = true;
  option (delphi.singleton)             = true;
  delphi.ObjectMeta  Meta               = 1;    // common delphi meta
  ForwardingMode     fwd_mode           = 2 [(gogoproto.jsontag) = "forwarding-mode,omitempty"];    // forwarding mode
  FeatureProfile     feature_profile    = 3 [(gogoproto.jsontag) = "feature-profile"]; // feature profile
  string             port_admin_state   = 4 [(gogoproto.jsontag) = "port-admin-state"];
  fixed64            mgmt_if_mac        = 5 [(gogoproto.jsontag) = "mgmt-if-mac"];
  MircoSegMode       micro_seg_mode     = 6 [(gogoproto.jsontag) = "micro-seg-mode"];
}
